File: practice/mathematics/algebra/game_of_rotation/game_of_rotation.py
#!/usr/bin/env python

n = int(raw_input().strip())
arr = map(int, raw_input().strip().split())

s  = sum(arr)
s1 = sum([(POS + 1) * arr[POS] for POS in xrange(n)])

part_sum = arr[0]
res = s1
for K in xrange(2, n + 1):
    sk = s1 - (K - 1) * s + n * part_sum
    part_sum += arr[K - 1]
    if sk > res:
        res = sk
print res

File: practice/mathematics/algebra/game_of_rotation/game_of_rotation.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>

using namespace std;

int main()
{
    int n;
    long long s, s1, sk, part_sum, res;
    
    cin >> n;
    
    vector<long long> arr(n);
    for (int pos = 0; pos < n; ++pos)
    {
        cin >> arr[pos];    
    }
    
    s = accumulate(arr.begin(), arr.end(), 0.0);
    
    s1 = 0;
    for (int pos = 0; pos < n; ++pos)
    {
        s1 += (pos + 1) * arr[pos];
    }
    
    part_sum = arr[0];
    res = s1;
    for (int k = 2; k < n + 1; ++k)
    {
        sk = s1 - (k - 1) * s + n * part_sum;
        part_sum += arr[k - 1];
        if (sk > res)
        {
            res = sk;
        }
    }
    
    cout << res << endl;
    
    return 0;
}

